File: cpp/tree/lowest_common_ancestor_binary_tree.cpp
// 二叉树的最近公共祖先

#include "stack"
#include "queue"
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() {}
    TreeNode(int val) : val(val) {}
    TreeNode(int val, TreeNode *left, TreeNode *right) : val(val), left(left), right(right) {}
};

class Solution {
public:

    // 如果 root 为空，或者 p q 任意一个等于 root，则直接返回
    // 递归遍历左右
    // 如果左右均不为空，则说明root就是最近公共祖先
    // 否则哪个不为空就返回哪个
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if (!root || root == p || root == q) {
            return root;
        }
        TreeNode* left = lowestCommonAncestor(root->left, p, q);
        TreeNode* right = lowestCommonAncestor(root->right, p, q);
        if (left && right) {
            return root;
        }
        return left == nullptr ? right : left;
    }
};

File: java/datastructure/tree/leetcode/LowestCommonAncestorBinaryTree.java
package datastructure.tree.leetcode;

/**
 * @author roseduan
 * @time 2022/2/27 9:53 上午
 * @description 二叉树的最近公共祖先
 */

public class LowestCommonAncestorBinaryTree {

    /**
     * 如果root为空，或者p q任意一个等于root，直接返回root
     * 分别遍历左右节点
     * 如果左右节点都不为空，则说明root就是最近的公共祖先
     * 如果有任意一个不为空，则直接返回
     */
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null ||root == p || root == q) {
            return root;
        }
        TreeNode l = lowestCommonAncestor(root.left, p, q);
        TreeNode r = lowestCommonAncestor(root.right, p, q);
        if (l != null && r != null) {
            return root;
        }
        return l == null ? r : l;
    }

    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
}
